autorest/adal/cli.go

```diff
@@ -47,13 +50,42 @@ func AzureCLIProfilePath() (string, error) {
 }
 
 // ToToken converts an AzureCLIToken to a Token
-func (t AzureCLIToken) ToToken() Token {
-	return Token{
+func (t AzureCLIToken) ToToken() (*Token, error) {
+	tokenExpirationDate, err := ParseAzureCLIExpirationDate(t.ExpiresOn)
+	if err != nil {
+		return nil, fmt.Errorf("Error parsing Token Expiration Date %q: %+v", t.ExpiresOn, err)
+	}
+
+	difference := tokenExpirationDate.Sub(expirationBase)
+	seconds := difference.Seconds()
+
+	token := Token{
 		AccessToken:  t.AccessToken,
 		Type:         t.TokenType,
-		ExpiresIn:    strconv.Itoa(t.ExpiresIn),
-		ExpiresOn:    t.ExpiresOn,
+		ExpiresIn:    "3600",
+		ExpiresOn:    strconv.Itoa(int(seconds)),
 		RefreshToken: t.RefreshToken,
 		Resource:     t.Resource,
 	}
+	return &token, nil
+}
+
+// ParseAzureCLIExpirationDate parses either a Azure CLI or CloudShell date into a time object
+func ParseAzureCLIExpirationDate(input string) (*time.Time, error) {
```

Comment:
Q: why do you return pointer to time.
time package uses value semantics by default, returning pointer causes one more allocation, moves object from stack to heap and creates work for garbage collector. as it is used only within a calling method it should return value.
this way it would stay on stack and get disposed when method finishes. no additional work for GC.

Reply:
I’ve opted for this approach since we’re returning nil for an invalid/unparseable date; given returning an empty object in addition to an error in the failure case feels misleading as a consumer of the API

Reply:
It seems like you use that pattern pretty religiously, @tombuildsstuff. I'm not really convinced either way, other than sorta general intuitiveness, can you build a case for why it's better to return a pointer? Personally, when I'm writing Go, I just treat retvals that accompany an error as garbage, it doesn't matter if the value is the default or memory with random bits. As you can imagine, in my workflow, changing the return type to a pointer has almost no consequence. What scenario am I missing where it's totally important? :)

Reply:
not particularly - I've only opted for this approach to force the consumer to handle the error case, rather than accidental comparing the values returned by the default object (it's more habit than anything else) - but totally happy to switch this out :)